README.md
# jepsen.history.sim

[![Clojars Project](https://img.shields.io/clojars/v/io.jepsen/history.sim.svg)](https://clojars.org/io.jepsen/history.sim)

When building a database checker, you need histories to check. These histories
should exhibit certain classes of anomalies, so that you can test whether the
checker is correct. They should be determistic, so you can build regression
tests. They should be large enough to serve as performance benchmarks. This
library generates those histories by simulating concurrent requests against
models of databases with assorted concurrency control schemes.

As a Clojure library, jepsen.history.sim provides histories to JVM programs
which ask for them. These histories are cached in your local tmpdir, so they're
fast on repeated access. Histories are read lazily from chunks, wrapped in a
jepsen.history.IHistory, so they work with parallel folds out of the box.

As an executable jar, this program can emit histories in EDN or JSON to stdout.
You can use this to benchmark checkers written in other languages.

## Library use

The main entry point for the library is `jepsen.history.sim`. Here's how to get a quick history from a snapshot isolated database:

```clj
=> (require '[jepsen.history.sim :as sim])
=> (pprint (sim/run {}))
{:concurrency 3,
 :limit 16,
 :db :si,
 :generator :list-append,
 :history
  [{:process 0,
   :type :invoke,
   :f :txn,
   :value [[:append 9 1] [:r 9 nil]],
   :index 0,
   :time -1}
  {:process 1,
   :type :invoke,
   :f :txn,
   :value [[:append 9 2]],
   :index 1,
   :time -1}
```

`run` takes and returns a test map. This map includes several parameters: the
db, the generator, concurrency, and so on. `run` fills in defaults if you don't
provide any. Once run, a test has a `:history` field with a Jepsen history of
operations.

Key distribution, transaction size, etc are all configurable; see
`jepsen.history.sim/run`'s docs. Let's say we wanted a prefix-consistent
history of 2 processes running 3 operations:

```clj
=> (->> (sim/run {:db :prefix, :concurrency 2, :limit 3}) :history (mapv prn))
{:index 0, :time -1, :type :invoke, :process 1, :f :txn, :value [[:append 9 1] [:r 9 nil]]}
{:index 1, :time -1, :type :invoke, :process 0, :f :txn, :value [[:append 9 2]]}
{:index 2, :time -1, :type :ok, :process 0, :f :txn, :value [[:append 9 2]]}
{:index 3, :time -1, :type :invoke, :process 0, :f :txn, :value [[:r 8 nil] [:r 9 nil]]}
{:index 4, :time -1, :type :ok, :process 0, :f :txn, :value [[:r 8 nil] [:r 9 [2]]]}
{:index 5, :time -1, :type :ok, :process 1, :f :txn, :value [[:append 9 1] [:r 9 [2 1]]]}
```

Runs are fully deterministic. Ask for the same parameters again and you'll get
an identical history. If you want a different random seed, you can ask for it:

```clj
=> (sim/run {:seed 3 :limit 1})
{:concurrency 3, :limit 1, :db :si, :generator :list-append, :seed 3, :history [{:index 0, :time -1, :type :invoke, :process 0, :f :txn, :value [[:r 9 nil]]} {:index 1, :time -1, :type :ok, :process 0, :f :txn, :value [[:r 9 nil]]}]}

=> (sim/run {:seed 4 :limit 1})
{:concurrency 3, :limit 1, :db :si, :generator :list-append, :seed 4, :history [{:index 0, :time -1, :type :invoke, :process 1, :f :txn, :value [[:append 9 1] [:r 4 nil]]} {:index 1, :time -1, :type :ok, :process 1, :f :txn, :value [[:append 9 1] [:r 4 nil]]}]}
```

Runs are cached to disk transparently, making subsequent requests for identical
parameters much, much faster. You can erase this cache with
`(jepsen.history.cache/clear!)`.

## CLI Use

To build a fat jar, first install a JDK, install leiningen, and run `lein
uberjar`. This will emit a jar in `target/history.sim-...-standalone.jar` with
an executable binary. Then try

```
java -jar target/history.sim-0.1.0-SNAPSHOT-standalone.jar history
```

This emits a short history of sixteen operations (16 invokes, 16 completions)
like so. These are list-append operations from a simulated snapshot isolated
database, and can be fed directly to Elle's list-append checker.

```clj
{:index 0, :time -1, :type :invoke, :process 0, :f :txn, :value [[:append 9 1] [:r 9 nil]]}
{:index 1, :time -1, :type :invoke, :process 1, :f :txn, :value [[:append 9 2]]}
{:index 2, :time -1, :type :invoke, :process 2, :f :txn, :value [[:r 8 nil] [:r 9 nil]]}
{:index 3, :time -1, :type :ok, :process 0, :f :txn, :value [[:append 9 1] [:r 9 [1]]]}
{:index 4, :time -1, :type :invoke, :process 0, :f :txn, :value [[:r 7 nil]]}
{:index 5, :time -1, :type :fail, :process 1, :f :txn, :value [[:append 9 2]]}
...
```

You can tune the number of operations with `--limit`, the concurrency with
`--concurrency`, the database with `--db`, and the output format with
`--format`. Here's a million operations by ten concurrent clients on a
serializable-snapshot-isolated database, printed as lines of JSON.

```
java -jar target/history.sim-0.1.0-SNAPSHOT-standalone.jar history --limit 1000000 --concurrency 10 --db prefix --format json
```

```json
{"index":0,"time":-1,"type":"invoke","process":7,"f":"txn","value":[["append",9,1],["r",9,null]]}
{"index":1,"time":-1,"type":"invoke","process":2,"f":"txn","value":[["append",9,2]]}
{"index":2,"time":-1,"type":"invoke","process":5,"f":"txn","value":[["r",8,null],["r",9,null]]}
{"index":3,"time":-1,"type":"invoke","process":4,"f":"txn","value":[["r",7,null]]}
{"index":4,"time":-1,"type":"invoke","process":6,"f":"txn","value":[["r",9,null],["r",9,null]]}
{"index":5,"time":-1,"type":"invoke","process":0,"f":"txn","value":[["r",9,null]]}
{"index":6,"time":-1,"type":"invoke","process":1,"f":"txn","value":[["append",6,1]]}
{"index":7,"time":-1,"type":"invoke","process":3,"f":"txn","value":[["append",8,1],["r",7,null]]}
{"index":8,"time":-1,"type":"invoke","process":8,"f":"txn","value":[["r",9,null],["r",6,null]]}
{"index":9,"time":-1,"type":"invoke","process":9,"f":"txn","value":[["append",8,2],["append",8,3]]}
```

This takes a while the first time, but run the command again and it'll start
printing immediately: the simulator will re-use its cached copy of the history
on disk. Takes about 35 seconds to finish printing all million ops on my aging
Xeon.

For full usage, run jepsen.history.sim without any arguments.

```
Usage: java -jar JAR_FILE [command] [args ...]

Commands:
  clear-cache  Clears the disk cache of histories.
  history      Emits a history to stdout

Options:

  -c, --concurrency NUM            3             How many concurrent clients?
      --db DB                      :si           What kind of database are we simulating?
      --format FORMAT              :edn          What format to emit operations in
  -g, --gen GENERATOR              :list-append  What kind of workload would you like to generate?
      --key-dist DISTRIBUTION      :exponential  Would you like an exponential or uniform distribution of keys in transactions?
      --key-dist-base NUMBER       2             Base for the exponential distribution of keys. 2 means each key is twice as frequent as the last.
      --key-count NUMBER           10            How many keys should transactions interact with at any given time?
      --min-txn-length NUMBER      1             How many micro-operations should transactions perform, at minimum?
      --max-txn-length NUMBER      4             How many micro-operations should transactions perform, at maximum?
      --max-writes-per-key NUMBER  32            How many writes should we try against each key before selecting a new one?
  -l, --limit NUM                  16            How many operations would you like to invoke?
      --seed NUM                   69            The random seed for this history
```

## DBs

Databases live in `jepsen.history.sim.db`. There are presently four
implementations:

- `brat`, which has no concurrency control whatsoever
- `prefix`, which is "snapshot isolation, without the first-committer-wins checkl"
- `si`, which is a by-the-book implementation of single-node snapshot isolation
- `ssi`, which makes SI serializable by promoting all writes to reads

## License

Copyright © 2023 Jepsen, LLC

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-12-07
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-12-07
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/jepsen.history.sim/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/jepsen.history.sim/compare/0.1.0...0.1.1


test/jepsen/history/sim_test.clj
(ns jepsen.history.sim-test
  (:refer-clojure :exclude [run!])
  (:require [clojure [pprint :refer [pprint]]
                     [test :refer :all]]
            [jepsen [history :as h]]
            [jepsen.history.sim :refer :all]
            [jepsen.history.sim [cache :as cache]]))

(deftest si-history-test
  (let [t (run {:db    :si
                :limit 4})]
    (is (= clojure.lang.PersistentVector
           (-> t :history first :value type)))
    (is (= {:max-txn-length 4,
           :concurrency 3,
           :limit 4,
           :key-count 10,
           :db :si,
           :max-writes-per-key 32,
           :key-dist :exponential,
           :history
           (h/history
           [{:process 0,
             :type :invoke,
             :f :txn,
             :value [[:append 9 1] [:r 9 nil] [:r 9 nil] [:append 9 2]],
             :index 0,
             :time -1}
            {:process 1,
             :type :invoke,
             :f :txn,
             :value [[:r 8 nil] [:r 9 nil] [:append 8 1]],
             :index 1,
             :time -1}
            {:process 2,
             :type :invoke,
             :f :txn,
             :value [[:append 9 3]],
             :index 2,
             :time -1}
            {:process 1,
             :type :ok,
             :f :txn,
             :value [[:r 8 nil] [:r 9 nil] [:append 8 1]],
             :index 3,
             :time -1}
            {:process 1,
             :type :invoke,
             :f :txn,
             :value [[:r 9 nil] [:append 8 2] [:r 9 nil]],
             :index 4,
             :time -1}
            {:process 2,
             :type :ok,
             :f :txn,
             :value [[:append 9 3]],
             :index 5,
             :time -1}
            {:process 1,
             :type :ok,
             :f :txn,
             :value [[:r 9 [3]] [:append 8 2] [:r 9 [3]]],
             :index 6,
             :time -1}
            {:process 0,
             :type :fail,
             :f :txn,
             :value [[:append 9 1] [:r 9 nil] [:r 9 nil] [:append 9 2]],
             :index 7,
             :time -1}]),
           :seed 69,
           :generator :list-append,
           :key-dist-base 2,
           :min-txn-length 1}
           t))))

(deftest brat-test
  (is (= (h/history
           [
            {:process 0,
            :type :invoke,
            :f :txn,
            :value [[:r 9 nil] [:append 9 1] [:append 6 1] [:append 9 2]],
            :index 0,
            :time -1}
           {:process 1,
            :type :invoke,
            :f :txn,
            :value [[:r 9 nil] [:r 6 nil] [:r 8 nil] [:append 9 3]],
            :index 1,
            :time -1}
           {:process 2,
            :type :invoke,
            :f :txn,
            :value [[:r 8 nil] [:r 8 nil]],
            :index 2,
            :time -1}
           {:process 1,
            :type :ok,
            :f :txn,
            :value [[:r 9 nil] [:r 6 nil] [:r 8 nil] [:append 9 [1 3]]],
            :index 3,
            :time -1}
           {:process 0,
            :type :ok,
            :f :txn,
            :value [[:r 9 nil] [:append 9 [1]] [:append 6 [1]] [:append 9 [1 3 2]]],
            :index 4,
            :time -1}
           {:process 2,
            :type :ok,
            :f :txn,
            :value [[:r 8 nil] [:r 8 nil]],
            :index 5,
            :time -1}
            ])
         (:history (run {:db :brat, :limit 3, :seed 123})))))

(defn ms
  "Helper for friendly nanos -> milliseconds"
  [nanos]
  (float (* 1e-6 nanos)))

(deftest cache-test
  ; Very helpful to lower this for debugging
  (with-redefs [cache/chunk-size 16384]
    (cache/clear!)
    (let [n    20000
          test (expand-test {:db :prefix, :limit n})]
      (is (nil? (cache/load test)))
      ; Uncached run
      (let [t0      (System/nanoTime)
            test0   (run! test)
            _       (is (= (* 2 n) (count (:history test0))))
            ; Now load cached copy
            t1      (System/nanoTime)
            test1   (run test)
            t2      (System/nanoTime)
            ; Comparing forces the test to load chunks
            _       (is (= test0 test1))
            t3      (System/nanoTime)]
        (println "Run took" (ms (- t1 t0))
                 "ms; load took" (ms (- t2 t1))
                 "ms; full read took" (ms (- t3 t1)) "ms")))))


test/jepsen/history/sim/db_test.clj
(ns jepsen.history.sim.db-test
  (:refer-clojure :exclude [read])
  (:require [clojure [pprint :refer [pprint]]
                     [test :refer :all]]
            [jepsen.history.sim.db :refer :all]))

(deftest si-write-skew-test
  ; We should be able to reproduce write skew with our SI DB
  (let [db (si-db)
        ; Start two transactions
        db (begin db 0)
        db (begin db 1)
        ; Both read intersecting keys
        [db r] (read db 0 :x)
        _ (is (= nil r))
        [db r] (read db 0 :y)
        _ (is (= nil r))
        [db r] (read db 1 :x)
        _ (is (= nil r))
        [db r] (read db 1 :y)
        _ (is (= nil r))
        ; And write disjoint keys
        [db _] (append db 0 :x 0)
        [db _] (append db 1 :y 1)
        ; Both may commit
        [db r] (commit db 0)
        _ (is (= true r))
        [db r] (commit db 1)
        _ (is (= true r))
        ; Which leaves the DB with both writes
        _ (is (= {:jepsen.history.sim.db/ts 2, :x [1 [0]], :y [2 [1]]} (:main db)))]
    ))


project.clj
(defproject io.jepsen/history.sim "0.1.1-SNAPSHOT"
  :description "Generates Jepsen histories from simulated databases, for testing safety checkers"
  :url "https://github.com/jepsen-io/history.sim"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[cheshire "5.12.0"]
                 [com.aphyr/bifurcan-clj "0.1.1"]
                 [io.jepsen/history "0.1.3"]
                 [org.clojure/clojure "1.11.1"]
                 [org.clojure/data.fressian "1.0.0"]
                 [org.clojure/data.generators "1.0.0"]
                 [org.clojure/tools.cli "1.0.219"]]
  :java-source-paths ["src/"]
  :main         jepsen.history.sim.cli
  :repl-options {:init-ns jepsen.history.sim}
  :test-selectors {:default (fn [m] (not (:perf m)))
                   :all         (fn [m] true)
                   :perf        :perf
                   :focus       :focus}
  :profiles {:uberjar {:aot :all}})


src/jepsen/history/sim.clj
(ns jepsen.history.sim
  "Top-level namespace for running simulations."
  (:refer-clojure :exclude [run!])
  (:require [clojure.data.generators :as data.gen]
            [jepsen.history.sim [cache :as cache]
                                [db :as db]
                                [gen :as gen]
                                [run :as run]])
  (:import (java.util Random)))

(defmacro with-rand
  "Takes a seed long and evaluates body with an RNG initialized to that seed."
  [seed & body]
  `(binding [data.gen/*rnd* (Random. ~seed)]
     ~@body))

(def test-defaults
  "Default parameters for tests"
  {:seed               69
   :concurrency        3
   :limit              16
   :db                 :si
   :generator          :list-append
   ; Generator params
   :key-dist           :exponential
   :key-dist-base      2
   :key-count          10
   :min-txn-length     1
   :max-txn-length     4
   :max-writes-per-key 32
   })


(defn expand-test
  "Expands a test with default options."
  [test]
  (merge test-defaults test))

(defn run!
  "Like run, but actually runs the simulator instead of potentially loading
  from cache. Saves to cache as a side effect."
  [test]
  (with-rand (:seed test)
    (let [gen-f (get gen/generators (:generator test))
          gen   (take (:limit test) (gen-f test))
          db-f  (get db/dbs (:db test))
          db    (db-f test)
          history (run/simulate (assoc test :generator gen :db db))]
      (-> (assoc test :history history)
          cache/write!
          cache/load))))

(defn run
  "Takes test parameters and returns a test with a completed :history field.
  Parameters are a map of:

    {; Database parameters
     :db        One of :si, :prefix, etc: the name of a DB

     ; Generator parameters
     :generator            The name of the generator to use, e.g. :list-append
     :key-count            Number of distinct keys at any point
     :min-txn-length       Minimum number of operations per txn
     :max-txn-length       Maximum number of operations per txn
     :max-writes-per-key   Maximum number of operations per key
     :concurrency          Number of concurrent clients
     :limit                Number of invocations to generate

     ; Randomness
     :seed                 A long: the random seed to use}

  This function will load an earlier run from disk cache, if possible."
  [test]
  (let [test (expand-test test)]
    (or (cache/load test)
        (run! test))))


src/jepsen/history/sim/db.clj
(ns jepsen.history.sim.db
  "Reference implementations of various databases."
  (:refer-clojure :exclude [read]))

; A Database executes transactions by applying read and append operations.
; Begin initiates a transaction for the given process. Commit attempts to
; finalize that transaction. All protocols here are functional, returning a new
; state of the database, followed by return values.
(defprotocol Database
  (begin [database process] "Begins a new transaction for the given process.
                            Returns the new state of the database.")
  (commit [database process] "Commits the transaction for the given process.
                             Returns a tuple: the new state of the database,
                             followed by true if the transaction committed,
                             false otherwise.")
  (abort [database process] "Aborts the transaction for the given process.
                            Returns a tuple: the new state of the database,
                            followed by true if the transaction aborted, which
                            should be always.")
  (read [database process key]
        "Reads the current state of a key. Takes the process performing the
        read. Returns [database', value]. value may be ::block, which indicates
        the given read can't be performed yet, but could be later.")
  (append [database process key element]
          "Appends the given element to the current value of the given key.
          Returns [database', element]. If element is ::block, the given
          append can't be performed yet, but could be later."))

; Brat performs no concurrency control. Aborts are not rolled back. You want
; G0s? Here you go.
(defrecord BratDB [state]
  Database
  (begin [db process] db)
  (commit [db process] [db true])
  (abort [db process] [db true])
  (read [db process k] [db (get state k)])
  (append [db process k element]
    (let [v (conj (get state k []) element)]
      [(BratDB. (assoc state k v)) v])))

(defn brat-db
  "Constructs a bratty database which performs no concurrency control and does
  not roll back aborted transactions."
  ([]
   (brat-db {}))
  ([opts]
   (BratDB. {})))

; Straightforward SI without first-committer-wins checks. This should give us
; prefix consistency, which is, to be clear, hot garbage.
;
; main      The main DB snapshot. A map of keys to [ts, value] pairs. A special
;           key, ::ts, identifies the logical timestamp for a snapshot's
;           state.
; snapshots a map of processes to snapshots taken for those txns.
; writes    a map of processes to maps of keys to values which were updated in
;           that txn. Used for conflict detection, and in-txn reads.
(defrecord PrefixDB [main snapshots writes]
  Database
  (begin [db process]
    (assoc db
           :snapshots (assoc snapshots  process (:main db))
           :writes    (assoc writes     process {})))

  (commit [db process]
    (let [start-ts  (::ts (get snapshots process))
          commit-ts (inc (::ts main))
          writes    (get writes process) ; Writes from this txn

          ; Iterate over writes in txn and apply them to main
          main' (reduce (fn [main [k v']]
                          ; Writes are always legal
                          (assoc main k [commit-ts v']))
                        (assoc main ::ts commit-ts)
                        writes)

          ; We'll use abort's logic to clean up transaction state for us.
          aborted-db (first (abort db process))]

        (if (= ::abort main')
          [aborted-db false]
          [(assoc aborted-db :main main') true])))

  (abort [db process]
    ; Just clear out our txn state.
    [(assoc db
            :snapshots (dissoc snapshots process)
            :writes    (dissoc writes process))
     true])

  (read [database process k]
    ; Read from in-txn writes first, fall back to snapshot
    [database (get-in writes [process k]
                      (second (get-in snapshots [process k])))])

  (append [database process k element]
    (let [v  (or (second (read database process k)) [])
          v' (conj v element)]
      [(assoc-in database [:writes process k] v')
       element])))

(defn prefix-db
  "Constructs a new prefix-isolated database"
  ([] (prefix-db {}))
  ([opts]
   (PrefixDB. {::ts 0} {} {})))

; A simple snapshot isolation model. Weirdly, this doesn't seem to emit
; G-singles or G-nonadjacents. Only G2-item. Must be a bug.
;
; main      The main DB snapshot. A map of keys to [ts, value] pairs. A special
;           key, ::ts, identifies the logical timestamp for a snapshot's
;           state.
; snapshots a map of processes to snapshots taken for those txns.
; writes    a map of processes to maps of keys to values which were updated in
;           that txn. Used for conflict detection, and in-txn reads.
(defrecord SIDB [main snapshots writes]
  Database
  (begin [db process]
    (assoc db
           :snapshots (assoc snapshots  process (:main db))
           :writes    (assoc writes     process {})))

  (commit [db process]
    (let [start-ts  (::ts (get snapshots process))
          commit-ts (inc (::ts main))
          writes    (get writes process) ; Writes from this txn

          ; Iterate over writes in txn and apply them to main, checking whether
          ; they were written between our start and commit timestamps.
          main' (reduce (fn [main [k v']]
                          (let [k-ts (first (get main k [0 nil]))]
                            ;(prn :start start-ts :k-ts k-ts :commit-ts commit-ts)
                            (if (< start-ts k-ts commit-ts)
                              ; Oh no, someone else updated!
                              (reduced ::abort)
                              ; Our write is legal
                              (assoc main k [commit-ts v']))))
                        (assoc main ::ts commit-ts)
                        writes)

          ; We'll use abort's logic to clean up transaction state for us.
          aborted-db (first (abort db process))]

        (if (= ::abort main')
          [aborted-db false]
          [(assoc aborted-db :main main') true])))

  (abort [db process]
    ; Just clear out our txn state.
    [(assoc db
            :snapshots (dissoc snapshots process)
            :writes    (dissoc writes process))
     true])

  (read [database process k]
    ; Read from in-txn writes first, fall back to snapshot
    [database (get-in writes [process k]
                      (second (get-in snapshots [process k])))])

  (append [database process k element]
    (let [v  (or (second (read database process k)) [])
          v' (conj v element)]
      [(assoc-in database [:writes process k] v')
       element])))

(defn si-db
  "Construct a fresh snapshot isolation DB."
  ([]
   (si-db {}))
  ([opts]
   (SIDB. {::ts 0} {} {})))

; Wrap a snapshot isolated DB in one that promotes all reads to writes, which
; ensures serializability.
(defrecord SSIDB [db]
  Database
  (begin [this process]
    (assoc this :db (begin db process)))

  (commit [this process]
    (let [[db' committed?] (commit db process)]
      [(assoc this :db db') committed?]))

  (abort [this process]
    (let [[db' ok?] (abort db process)]
      [(assoc this :db db') ok?]))

  (read [this process k]
    (let [[db' v] (read db process k)
          ; Sneak into the underlying SI DB and add a no-op write
          db' (assoc-in db' [:writes process k] v)]
      [(assoc this :db db') v]))

  (append [this process k element]
    (let [[db' r] (append db process k element)]
      [(assoc this :db db') r])))

(defn ssi-db
  "Construct a fresh snapshot serializable isolation database."
  ([] (ssi-db {}))
  ([opts]
  (SSIDB. (si-db))))

(def dbs
  "A map of standard DB types to DB constructor functions."
  {:brat   brat-db
   :prefix prefix-db
   :si     si-db
   :ssi    ssi-db})


src/jepsen/history/sim/run.clj
(ns jepsen.history.sim.run
  "Takes databases and generators and runs them, producing histories."
  (:require [bifurcan-clj [core :as b]
                          [map :as bm]
                          [set :as bs]]
            [clojure.data.generators :as data.gen]
            [jepsen [history :as h]]
            [jepsen.history.sim [db :as db]])
  (:import (java.util Random)))

(defn brand-nth
  "Rand-nth on a Bifurcan collection. Returns nil on empty colls."
  [xs]
  (let [n (b/size xs)]
    (when (< 0 n)
      (let [i (mod (data.gen/long) n)]
        (b/nth xs i)))))

(defn simulate
  "Takes a map with the following keys:

  {:concurrency   The number of concurrent clients
   :db            The database to execute against
   :generator     A sequence of invocations to perform}

  Launches simulated clients with the given concurrency, which draw operations
  from the generator and apply them to the DB. Returns a Jepsen IHistory.

  Deterministically random; uses data.generators/*rnd*."
  ; We track that client state in a map of processes to {:state state, :applied
  ; [& micro-ops], :pending [& micro-ops]}, corresponding to those operations
  ; we've completed, and those which have yet to be applied. At each moment in
  ; time, we pick a random process to step forward, and have that client
  ; perform one of its pending micro-operations against the database, creating
  ; a corresponding micro-op in that process' :applied. When a client has
  ; applied all of its operations, it can try to commit. The results of that
  ; commit are used to construct a completion operation, which is appended to
  ; the history.
  [test]
  (loop [gen      (:generator test) ; Remaining sequence of invocations
         db       (:db test)        ; DB state
         ; The history is initially empty
         history  []
         ; Bifurcan map of all busy process IDs to client state
         busy     bm/empty
         ; Bifurcan set of free processes
         free     (bs/from (range (:concurrency test)))]
    (cond ; A process is idle.
          (pos? (b/size free))
          (let [process (brand-nth free)]
            ; Can we begin an operation?
            (if-let [op (first gen)]
              (let [op (h/op (assoc op
                                    :process process
                                    :index   (long (count history))))]
                ; We have an invocation. Journal it to the history, and set up
                ; the client.
                (recur (next gen)
                       db
                       (conj history op)
                       (bm/put busy process {:state    :beginning
                                             :invoke   op
                                             :pending  (:value op)
                                             :applied  []})
                       (bs/remove free process)))
              ; Out of operations; shut down this process.
              (recur gen
                     db
                     history
                     busy
                     (bs/remove free process))))

          ; No processes idle, and one is busy. Pick a random busy process
          ; and advance it.
          (pos? (b/size busy))
          (let [pair    (brand-nth busy)
                process (bm/key pair)
                {:keys [state invoke pending applied] :as pstate}
                (bm/value pair)]
            (case state
              ; We need to begin a transaction on the DB, then we can
              ; transition to in-txn.
              :beginning
              (recur gen
                     (db/begin db process)
                     history
                     (bm/put busy process (assoc pstate :state :in-txn))
                     free)

              ; We're applying operations from the transaction.
              :in-txn
              (if-not (seq pending)
                ; Out of ops; commit
                (let [[db' committed?] (db/commit db process)
                      op' (assoc invoke
                                 :index (count history)
                                 :type (if committed? :ok :fail)
                                 :value (if committed?
                                          applied
                                          (:value invoke)))]
                  (recur gen
                         db'
                         (conj history op')
                         (bm/remove busy process)
                         (bs/add free process)))
                ; Apply an op
                (let [[f k v]  (first pending)
                      [db' v'] (case f
                                 :r      (db/read   db process k)
                                 :append (db/append db process k v))]
                  (recur gen
                         db'
                         history
                         (bm/put busy process
                                 {:state   :in-txn
                                  :invoke  invoke
                                  :pending (next pending)
                                  :applied (conj applied [f k v'])})
                         free)))))

          ; No processes free or busy. Done!
          true history)))


src/jepsen/history/sim/cli.clj
(ns jepsen.history.sim.cli
  "Command-line interface for history.sim."
  (:require [cheshire.core :as json]
            [clojure.tools.cli :as cli]
            [jepsen.history.sim :as sim :refer [test-defaults]]
            [jepsen.history.sim [cache :as cache]
                                [db :as db]
                                [gen :as gen]])
  (:gen-class))


(def opt-spec
  "tools.cli options"
  [
   ["-c" "--concurrency NUM" "How many concurrent clients?"
    :default 3
    :parse-fn parse-long
    :validate [pos? "must be positive"]]

   [nil "--db DB" "What kind of database are we simulating?"
    :default :si
    :parse-fn keyword
    :validate [db/dbs (str "must be one of " (sort (keys db/dbs)))]]

   [nil "--format FORMAT" "What format to emit operations in"
    :default :edn
    :parse-fn keyword
    :validate [#{:json :edn} "must be edn or json"]]

   ["-g" "--gen GENERATOR" "What kind of workload would you like to generate?"
    :default :list-append
    :parse-fn keyword
    :validate [gen/generators
               (str "must be one of " (sort (keys gen/generators)))]]

   [nil "--key-dist DISTRIBUTION" "Would you like an exponential or uniform distribution of keys in transactions?"
    :default (:key-dist test-defaults)
    :parse-fn keyword
    :validate [#{:exponential :uniform} "must be either exponential or uniform"]]

   [nil "--key-dist-base NUMBER" "Base for the exponential distribution of keys. 2 means each key is twice as frequent as the last."
    :default (:key-dist-base test-defaults)
    :parse-fn parse-double]

   [nil "--key-count NUMBER" "How many keys should transactions interact with at any given time?"
    :default (:key-count test-defaults)
    :parse-fn parse-long
    :validate [pos? "must be positive"]]

   [nil "--min-txn-length NUMBER" "How many micro-operations should transactions perform, at minimum?"
    :default (:min-txn-length test-defaults)
    :parse-fn parse-long
    :validate [(complement neg?) "must not be negative"]]

   [nil "--max-txn-length NUMBER" "How many micro-operations should transactions perform, at maximum?"
    :default (:max-txn-length test-defaults)
    :parse-fn parse-long
    :validate [(complement neg?) "must not be negative"]]

   [nil "--max-writes-per-key NUMBER" "How many writes should we try against each key before selecting a new one?"
    :default (:max-writes-per-key test-defaults)
    :parse-fn parse-long
    :validate [pos? "must be positive"]]

   ["-l" "--limit NUM" "How many operations would you like to invoke?"
    :default 16
    :parse-fn parse-long
    :validate [(complement neg?) "must be non-negative"]]

   [nil "--seed NUM" "The random seed for this history"
    :default 69
    :parse-fn parse-long]
   ])

(defn history!
  "Generates a history and prints it to stdout."
  [options]
  (let [test (sim/run options)
        printer (case (:format options)
                  :json  (comp println json/generate-string)
                  :edn   prn)]
    (run! printer (:history test))))

(defn usage!
  "Prints usage to STDOUT."
  [summary]
  (do (println "Usage: java -jar JAR_FILE [command] [args ...]")
      (println)
      (println "Commands:")
      (println "  clear-cache  Clears the disk cache of histories.")
      (println "  history      Emits a history to stdout")
      (println)
      (println "Options:")
      (println)
      (println summary)))

(defn -main
  ([]
   (usage! (:summary (cli/parse-opts [] opt-spec)))
   (System/exit 0))
  ([cmd & args]
   (try
     (let [{:keys [options arguments summary errors]}
           (cli/parse-opts args opt-spec)]
       (when (seq errors)
         (doseq [e errors]
           (println e))
         (System/exit 2))

       (case cmd
         "clear-cache" (cache/clear!)

         "history" (history! options)

         (do (usage! summary)
             (System/exit 1))))

     (catch Throwable t
       (println t)
       (.printStackTrace t)
       (System/exit 1)))))
